engine/common/follower/engine.go

```diff
@@ -368,8 +369,12 @@ func (e *Engine) processBlockAndDescendants(ctx context.Context, proposal *messa
 	log.Info().Msg("forwarding block proposal to hotstuff")
 
 	// submit the model to follower for processing
-	e.follower.SubmitProposal(header, parent.View)
-
+	select {
+	case <-e.follower.SubmitProposal(header, parent.View):
```

Comment:
this will wait for every single proposal, but we only want to wait if the block is one of the range response. Right?

Reply:
We don't want to for every single proposal, because if follower engine didn't receive the next proposal, it doesn't need to wait. 

We only want to wait if we have received a range response, because after processing a block, we know the next block has already received, and we want to wait until this block to be processed before processing the next.

In order to do that, you will need to send the entire BlockResponse to the follower engine, instead of [processing every single `SyncBlock`](https://github.com/onflow/flow-go/blob/master/engine/common/synchronization/engine.go#L305-L318)

When follower engine process a range of block, it could call onBlockProposal to process each proposal as usual, but it will also wait for the callback, whereas the original onBlockProposal will not wait for the callback.